ncm-iptables/src/main/perl/iptables.pm

```diff
@@ -739,43 +697,41 @@ sub WriteFile {
 #   OUTPUT: $?        - 0 the rules are equal,
 #                     - 1 one, or the two rules, is empty or is not an
 #                         hash tables, or the rules are different.
-##########################################################################
 sub cmp_rules {
-    my ($rule1, $rule2) = @_;
-    my ($field);
-    my (@fields1, @fields2);
+    my ($self, $rule1, $rule2) = @_;
+
+    $self->debug(5, "Entering method cmp_rules");
 
     # Check parameters.
-    if ( ! defined $rule1 || ref($rule1) !~ /^HASH/ ) {
-	$? = 1;
-	$@ = "first rule is not an hash table";
-	return $?;
+    if (!defined $rule1 || ref($rule1) !~ /^HASH/) {
+        $? = 1;
+        $@ = "first rule is not an hash table";
+        return $?;
     }
-    if ( ! defined $rule2 || ref($rule2) !~ /^HASH/ ) {
-	$? = 1;
-	$@ = "second rule is not an hash table";
-	return $?;
+    if (!defined $rule2 || ref($rule2) !~ /^HASH/) {
+        $? = 1;
+        $@ = "second rule is not an hash table";
+        return $?;
     }
 
     $? = 1;
     $@ = "rule is not in the list";
 
-    @fields1 = keys %{$rule1};
-    @fields2 = keys %{$rule2};
+    my @fields1 = keys %{$rule1};
+    my @fields2 = keys %{$rule2};
 
-    return $? if ( scalar(@fields1) <= 0 && scalar(@fields2) >  0 );
-    return $? if ( scalar(@fields1) >  0 && scalar(@fields2) <= 0 );
+    return $? if (scalar(@fields1) <= 0 && scalar(@fields2) >  0);
```

Comment:
I'm now fairly sure this function doesn't actually work as documented, or at least the unit tests I have written seem to prove it doesn't.